### PR TITLE
feat(server): auto index all workspaces to indexer

### PR DESCRIPTION
**Pull request by fengmk2**
close CLOUD-207

## Summary by CodeRabbit

- **New Features**
  - Introduced automated periodic indexing of workspaces with a new job type and a scheduled cron job running every 30 seconds.
  - Added a unique sequential identifier (`sid`) and an "indexed" flag to workspaces to track indexing status.
- **Improvements**
  - Enhanced workspace indexing to handle missing workspaces and snapshots distinctly and selectively index documents.
  - Added ability to query workspaces after a given identifier with result limits.
- **Bug Fixes**
  - Improved error handling and logging during workspace indexing operations.
- **Tests**
  - Expanded test coverage for workspace indexing and auto-indexing, including scheduling and edge cases.
- **Chores**
  - Updated data models and schema to support new workspace fields and indexing features.
  - Enhanced mock data utilities to allow custom timestamps.
  - Improved type safety and flexibility in document snapshot retrieval.

### CODE REVIEW

**Comment by fengmk2** [comment]
* **#12245** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/12245?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#12205** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/12205?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/12205?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#11937** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11937?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#11955** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11955?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#11835** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11835?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#11836](https://github.com/toeverything/AFFiNE/pull/11836) <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/11836?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#12189** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/12189?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `canary`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/toeverything?ref=merge-queue-instructions-in-stack-comment&prId=12205).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/backend/server/src/plugins/indexer/config.ts [hunk 1]
```diff
@@ -18,6 +18,11 @@ declare global {
       username: string;
       password: string;
     };
+    indexerSync: {
```
Comment (forehalo) [inline packages/backend/server/src/plugins/indexer/config.ts]: make it sub config of indexer

File: packages/backend/server/src/plugins/indexer/config.ts [hunk 2]
```diff
@@ -58,4 +62,14 @@ defineModuleConfig('indexer', {
     default: '',
     env: ['AFFINE_INDEXER_SEARCH_PASSWORD', 'string'],
   },
+  'autoIndexState.lastIndexedWorkspaceSid': {
```
Comment (forehalo) [inline packages/backend/server/src/plugins/indexer/config.ts]: this is not suitable to be put in config

File: packages/backend/server/src/plugins/indexer/job.ts [hunk 3]
```diff
@@ -107,4 +112,55 @@ export class IndexerJob {
...
+      this.config.indexer.autoIndexState.limit
+    );
+    if (workspaces.length === 0) {
+      return;
+    }
+    for (const workspace of workspaces) {
+      await this.queue.add(
+        'indexer.indexWorkspace',
+        { workspaceId: workspace.id },
+        { jobId: workspace.id }
```
Comment (forehalo) [inline packages/backend/server/src/plugins/indexer/job.ts:130]: add some prefix

File: packages/backend/server/src/plugins/indexer/config.ts (hunk 2 above)
Comment (fengmk2) [reply packages/backend/server/src/plugins/indexer/config.ts]: A new `indexed` field is added to the workspaces table to indicate whether the index has been created. And we can also reset it to rebuild the index

File: packages/backend/server/src/plugins/indexer/job.ts [hunk 4]
```diff
@@ -107,4 +125,49 @@ export class IndexerJob {
...
+      );
+      addedCount++;
+    }
+    const nextSid = workspaces[workspaces.length - 1].sid;
+    this.logger.log(
+      `Auto added ${addedCount} workspaces to queue, lastIndexedWorkspaceSid: ${startSid} -> ${nextSid}`
+    );
+
+    // update the lastIndexedWorkspaceSid in the payload and repeat the job after 30 seconds
+    payload.lastIndexedWorkspaceSid = nextSid;
```
Comment (fengmk2) [inline packages/backend/server/src/plugins/indexer/job.ts:172]: @forehalo  Reconstitute a pure job to pass the data, and add the `indexed` property to determine whether to update the index fully

File: packages/backend/server/src/models/workspace.ts [hunk 5]
```diff
@@ -76,6 +77,18 @@ export class WorkspaceModel extends BaseModel {
     });
   }
 
+  async findManyBySid(sid: number, limit: number) {
```
Comment (forehalo) [inline packages/backend/server/src/models/workspace.ts]: listAfterSid

File: packages/backend/server/src/plugins/indexer/job.ts [hunk 6]
```diff
@@ -107,4 +125,66 @@ export class IndexerJob {
...
+    return JOB_SIGNAL.Repeat;
+  }
+
+  @Cron(CronExpression.EVERY_30_SECONDS)
```
Comment (forehalo) [inline packages/backend/server/src/plugins/indexer/job.ts:176]: cron is not required because of handler returns `REPEAT`

Comment (fengmk2) [reply packages/backend/server/src/plugins/indexer/job.ts:176]: This scheduled task triggers the first task and ensures that if the task is deleted, it can be automatically resumed.

**Review by forehalo — APPROVED** [review]
